Investigate intermittent test failures

We have seen a few test failures in `ManagedLeaderLatchCreatorTest` on recent builds. It is not always the same test, however the error seems to have the same cause:

```
Error:  Tests run: 12, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.618 s <<< FAILURE! - in org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest
Error:  org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldNotRegisterResources_IfDisabled  Time elapsed: 0.023 s  <<< ERROR!
org.kiwiproject.curator.leader.exception.ManagedLeaderLatchException: Error starting leader latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldNotRegisterResources_IfDisabled(ManagedLeaderLatchCreatorTest.java:239)
Caused by: org.kiwiproject.curator.leader.exception.ManagedLeaderLatchException: Unable to create latch path: /kiwi/leader-latch/test-service/leader-latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldNotRegisterResources_IfDisabled(ManagedLeaderLatchCreatorTest.java:239)
Caused by: org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists for /kiwi/leader-latch/test-service/leader-latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldNotRegisterResources_IfDisabled(ManagedLeaderLatchCreatorTest.java:239)
```

See https://github.com/kiwiproject/dropwizard-leader-latch/runs/1407650945#step:6:2270

This doesn't make much sense since we have an `@AfterEach` method that deletes the node `/kiwi/leader-latch` _and all its child znodes_. Here is the current method:

```java
@AfterEach
void tearDown() throws Exception {
    var rootPath = "/kiwi/leader-latch";
    var stat = client.checkExists().forPath(rootPath);
    if (nonNull(stat)) {
        client.delete().deletingChildrenIfNeeded().forPath(rootPath);
    }
}
```
We did not see exceptions in the `tearDown` method, so maybe the `checkExists` call is failing? Or maybe the deletion silently fails? Or, although I think the delete call is _synchronous_, maybe it actually isn't?

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
@@ -5,7 +5,7 @@ import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 import static org.kiwiproject.collect.KiwiLists.first;
 import static org.kiwiproject.collect.KiwiLists.second;
-import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursively;
+import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursivelyIfExists;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.ArgumentMatchers.eq;
@@ -78,7 +78,7 @@ class ManagedLeaderLatchCreatorTest {
     void tearDown() throws Exception {
         var rootPath = "/kiwi/leader-latch";
 
-        deleteRecursively(client, rootPath);
+        deleteRecursivelyIfExists(client, rootPath);
     }
 
     @Test

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
@@ -4,6 +4,7 @@ import static java.util.Objects.nonNull;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
+import static org.awaitility.Durations.TWO_SECONDS;
 import static org.kiwiproject.collect.KiwiLists.first;
 import static org.kiwiproject.collect.KiwiLists.second;
 import static org.mockito.ArgumentMatchers.any;
@@ -45,6 +46,7 @@ import java.util.concurrent.atomic.AtomicBoolean;
 
 @DisplayName("ManagedLeaderLatchCreator")
 @ExtendWith(SoftAssertionsExtension.class)
+@Slf4j
 class ManagedLeaderLatchCreatorTest {
 
     @RegisterExtension
@@ -76,10 +78,26 @@ class ManagedLeaderLatchCreatorTest {
     @AfterEach
     void tearDown() throws Exception {
         var rootPath = "/kiwi/leader-latch";
-        var stat = client.checkExists().forPath(rootPath);
-        if (nonNull(stat)) {
-            client.delete().deletingChildrenIfNeeded().forPath(rootPath);
+
+        if (pathExists(rootPath)) {
+            LOG.debug("Path {} exists, attempting to delete it", rootPath);
+            client.delete().guaranteed().deletingChildrenIfNeeded().forPath(rootPath);
+
+            // In GitHub we have intermittent test failures caused by NodeExistsException thrown in setUp.
+            // The following attempts to wait and see if it gets deleted. Also, added guaranteed() in the
+            // above code that attempts the deletion, which causes Curator to attempt background deletes.
+            // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
+            if (pathExists(rootPath)) {
+                LOG.warn("Path {} still exists; wait up to two seconds for it to be deleted", rootPath);
+                await().atMost(TWO_SECONDS).until(() -> !pathExists(rootPath));
+            }
         }
+    }
+
+    private boolean pathExists(String rootPath) throws Exception {
+        var stat = client.checkExists().forPath(rootPath);
+
+        return nonNull(stat);
     }
 
     @Test

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
@@ -81,11 +81,10 @@ class ManagedLeaderLatchCreatorTest {
 
         if (pathExists(rootPath)) {
             LOG.debug("Path {} exists, attempting to delete it", rootPath);
-            client.delete().guaranteed().deletingChildrenIfNeeded().forPath(rootPath);
+            client.delete().deletingChildrenIfNeeded().forPath(rootPath);
 
             // In GitHub we have intermittent test failures caused by NodeExistsException thrown in setUp.
-            // The following attempts to wait and see if it gets deleted. Also, added guaranteed() in the
-            // above code that attempts the deletion, which causes Curator to attempt background deletes.
+            // The following attempts to wait and see if it gets deleted.
             // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
             if (pathExists(rootPath)) {
                 LOG.warn("Path {} still exists; wait up to five seconds for it to be deleted", rootPath);

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
@@ -4,7 +4,7 @@ import static java.util.Objects.nonNull;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
-import static org.awaitility.Durations.TWO_SECONDS;
+import static org.awaitility.Durations.FIVE_SECONDS;
 import static org.kiwiproject.collect.KiwiLists.first;
 import static org.kiwiproject.collect.KiwiLists.second;
 import static org.mockito.ArgumentMatchers.any;
@@ -88,8 +88,8 @@ class ManagedLeaderLatchCreatorTest {
             // above code that attempts the deletion, which causes Curator to attempt background deletes.
             // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
             if (pathExists(rootPath)) {
-                LOG.warn("Path {} still exists; wait up to two seconds for it to be deleted", rootPath);
-                await().atMost(TWO_SECONDS).until(() -> !pathExists(rootPath));
+                LOG.warn("Path {} still exists; wait up to five seconds for it to be deleted", rootPath);
+                await().atMost(FIVE_SECONDS).until(() -> !pathExists(rootPath));
             }
         }
     }

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
@@ -1,12 +1,11 @@
 package org.kiwiproject.curator.leader;
 
-import static java.util.Objects.nonNull;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
-import static org.awaitility.Durations.FIVE_SECONDS;
 import static org.kiwiproject.collect.KiwiLists.first;
 import static org.kiwiproject.collect.KiwiLists.second;
+import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursively;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.ArgumentMatchers.eq;
@@ -79,24 +78,7 @@ class ManagedLeaderLatchCreatorTest {
     void tearDown() throws Exception {
         var rootPath = "/kiwi/leader-latch";
 
-        if (pathExists(rootPath)) {
-            LOG.debug("Path {} exists, attempting to delete it", rootPath);
-            client.delete().deletingChildrenIfNeeded().forPath(rootPath);
-
-            // In GitHub we have intermittent test failures caused by NodeExistsException thrown in setUp.
-            // The following attempts to wait and see if it gets deleted.
-            // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
-            if (pathExists(rootPath)) {
-                LOG.warn("Path {} still exists; wait up to five seconds for it to be deleted", rootPath);
-                await().atMost(FIVE_SECONDS).until(() -> !pathExists(rootPath));
-            }
-        }
-    }
-
-    private boolean pathExists(String rootPath) throws Exception {
-        var stat = client.checkExists().forPath(rootPath);
-
-        return nonNull(stat);
+        deleteRecursively(client, rootPath);
     }
 
     @Test

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchTest.java
@@ -5,6 +5,7 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 import static org.awaitility.Durations.FIVE_SECONDS;
+import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursively;
 import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
 
@@ -64,7 +65,7 @@ class ManagedLeaderLatchTest {
         leaderLatch2 = new ManagedLeaderLatch(client, "id-67890", "test-service", leaderListener2);
 
         if (exists(leaderLatch1.getLatchPath())) {
-            client.delete().deletingChildrenIfNeeded().forPath(leaderLatch1.getLatchPath());
+            deleteRecursively(client, leaderLatch1.getLatchPath());
         }
     }
 
@@ -368,8 +369,9 @@ class ManagedLeaderLatchTest {
     }
 
     @Test
-    void shouldCallActionAsynchronously_WhenWhenIsLeader() throws Exception {
+    void shouldCallActionAsynchronously_WhenIsLeader() throws Exception {
         startAndAwait(leaderLatch1);
+        assertThat(leaderLatch1.hasLeadership()).isTrue();
 
         var called = new AtomicBoolean();
         Optional<CompletableFuture<Void>> result = leaderLatch1.whenLeaderAsync(() -> called.set(true));
@@ -381,7 +383,7 @@ class ManagedLeaderLatchTest {
     }
 
     @Test
-    void shouldNotCallActionAsynchronously_WhenWhenIsNotLeader() throws Exception {
+    void shouldNotCallActionAsynchronously_WhenIsNotLeader() throws Exception {
         startAndAwait(leaderLatch1);
         startAndAwait(leaderLatch2);
         assertThat(leaderLatch1.hasLeadership()).isTrue();

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchTest.java
@@ -5,7 +5,7 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 import static org.awaitility.Durations.FIVE_SECONDS;
-import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursively;
+import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursivelyIfExists;
 import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
 
@@ -65,7 +65,7 @@ class ManagedLeaderLatchTest {
         leaderLatch2 = new ManagedLeaderLatch(client, "id-67890", "test-service", leaderListener2);
 
         if (exists(leaderLatch1.getLatchPath())) {
-            deleteRecursively(client, leaderLatch1.getLatchPath());
+            deleteRecursivelyIfExists(client, leaderLatch1.getLatchPath());
         }
     }
 

--- a/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
+++ b/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
@@ -17,7 +17,7 @@ public class CuratorTestHelpers {
             LOG.debug("Path {} exists, attempting to delete it", rootPath);
             client.delete().deletingChildrenIfNeeded().forPath(rootPath);
 
-            // In GitHub we have intermittent test failures caused by NodeExistsException thrown in setUp.
+            // In GitHub we have seen various intermittent test failures caused by NodeExistsException.
             // The following attempts to wait and see if it gets deleted.
             // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
             if (pathExists(client, rootPath)) {

--- a/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
+++ b/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
@@ -1,0 +1,35 @@
+package org.kiwiproject.curator.leader.util;
+
+import static java.util.Objects.nonNull;
+import static org.awaitility.Awaitility.await;
+import static org.awaitility.Durations.FIVE_SECONDS;
+
+import lombok.experimental.UtilityClass;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.CuratorFramework;
+
+@UtilityClass
+@Slf4j
+public class CuratorTestHelpers {
+
+    public static void deleteRecursively(CuratorFramework client, String rootPath) throws Exception {
+        if (pathExists(client, rootPath)) {
+            LOG.debug("Path {} exists, attempting to delete it", rootPath);
+            client.delete().deletingChildrenIfNeeded().forPath(rootPath);
+
+            // In GitHub we have intermittent test failures caused by NodeExistsException thrown in setUp.
+            // The following attempts to wait and see if it gets deleted.
+            // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
+            if (pathExists(client, rootPath)) {
+                LOG.warn("Path {} still exists; wait up to five seconds for it to be deleted", rootPath);
+                await().atMost(FIVE_SECONDS).until(() -> !pathExists(client, rootPath));
+            }
+        }
+    }
+
+    public static boolean pathExists(CuratorFramework client, String rootPath) throws Exception {
+        var stat = client.checkExists().forPath(rootPath);
+
+        return nonNull(stat);
+    }
+}

--- a/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
+++ b/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
@@ -1,34 +1,59 @@
 package org.kiwiproject.curator.leader.util;
 
 import static java.util.Objects.nonNull;
-import static org.awaitility.Awaitility.await;
-import static org.awaitility.Durations.FIVE_SECONDS;
 
 import lombok.experimental.UtilityClass;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.curator.framework.CuratorFramework;
+import org.kiwiproject.retry.SimpleRetryer;
+import org.slf4j.event.Level;
+
+import java.util.concurrent.TimeUnit;
 
 @UtilityClass
 @Slf4j
 public class CuratorTestHelpers {
 
-    public static void deleteRecursively(CuratorFramework client, String rootPath) throws Exception {
-        if (pathExists(client, rootPath)) {
-            LOG.debug("Path {} exists, attempting to delete it", rootPath);
-            client.delete().deletingChildrenIfNeeded().forPath(rootPath);
-
-            // In GitHub we have seen various intermittent test failures caused by NodeExistsException.
-            // The following attempts to wait and see if it gets deleted.
-            // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
-            if (pathExists(client, rootPath)) {
-                LOG.warn("Path {} still exists; wait up to five seconds for it to be deleted", rootPath);
-                await().atMost(FIVE_SECONDS).until(() -> !pathExists(client, rootPath));
-            }
+    public static void deleteRecursivelyIfExists(CuratorFramework client, String path) throws Exception {
+        if (pathExists(client, path)) {
+            deleteRecursively(client, path);
         }
     }
 
-    public static boolean pathExists(CuratorFramework client, String rootPath) throws Exception {
-        var stat = client.checkExists().forPath(rootPath);
+    public static void deleteRecursively(CuratorFramework client, String path) {
+        // In GitHub we have seen various intermittent test failures caused by NodeExistsException.
+        // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
+
+        // The following attempts to delete a path and any children. If the path still exists after
+        // attempting to delete it, try again up to a max of 5 attempts.
+
+        var retryer = SimpleRetryer.builder()
+                .commonType("delete path: " + path)
+                .retryDelayTime(1)
+                .retryDelayUnit(TimeUnit.SECONDS)
+                .maxAttempts(5)
+                .logLevelForSubsequentAttempts(Level.WARN)
+                .build();
+
+        // NOTE: returning null forces a retry unless max attempts have been reached
+        retryer.tryGetObject(() -> {
+            try {
+                deletePathAndChildren(client, path);
+
+                return pathExists(client, path) ? null : path;
+            } catch (Exception e) {
+                LOG.warn("Error deleting path recursively: {}", path, e);
+                return null;
+            }
+        });
+    }
+
+    private static void deletePathAndChildren(CuratorFramework client, String path) throws Exception {
+        client.delete().deletingChildrenIfNeeded().forPath(path);
+    }
+
+    public static boolean pathExists(CuratorFramework client, String path) throws Exception {
+        var stat = client.checkExists().forPath(path);
 
         return nonNull(stat);
     }

